### PR TITLE
Add update for tournament

### PR DESCRIPTION
**Pull request by Fastjur**
Signed-off-by: Jurriaan Den Toonder <jur.den.toonder@gmail.com>

### CODE REVIEW

File: src/main/java/ch/wisv/areafiftylan/web/tournament/controller/TournamentController.java [hunk 1]
```diff
@@ -42,6 +43,29 @@ public TournamentController(TournamentService tournamentService) {
         return createResponseEntity(HttpStatus.CREATED, "Tournament added.", tournament);
     }
 
+    @PostMapping("/{id}")
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/web/tournament/controller/TournamentController.java]: As we're editing, this should be a PUT mapping

File: src/main/java/ch/wisv/areafiftylan/web/tournament/controller/TournamentController.java [hunk 2]
```diff
@@ -42,6 +43,29 @@ public TournamentController(TournamentService tournamentService) {
...
+    @PreAuthorize("hasRole('COMMITTEE')")
+    ResponseEntity<?> editTournament(@RequestBody Tournament tournament) {
+        Tournament dbTournament = tournamentService.getTournaments()
+                .stream()
+                .filter(tournament1 -> tournament1.getId().equals(tournament.getId()))
+                .findFirst()
+                .orElseThrow(TournamentNotFoundException::new);
+
+        // This is a bit double, but it ensures we update an existing tournament
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/web/tournament/controller/TournamentController.java]: Is there a field we're not updating here? If you're literally replacing them all, you can just replace it with the new responsebody object. Also, this is typical service logic, shouldn't be in the controller

File: src/test/java/ch/wisv/areafiftylan/integration/WebTournamentIntegrationTest.java [hunk 3]
```diff
@@ -212,4 +234,67 @@ public void testDeleteAllTournamentsAsCommitteeMember() {
...
+            .header(getXAuthTokenHeaderForUser(user))
+            .body(updateTournament)
+            .contentType(ContentType.JSON)
+        .when()
+            .post(TOURNAMENT_ENDPOINT + tournament.getId())
+        .then()
+            .statusCode(HttpStatus.SC_FORBIDDEN);
+
+        assertEquals(1, tournamentRepository.findAll().size());
+        /* toString() is used to force it to check the values, rather than the objects (somehow this failed)
```
Comment (skamoen) [inline src/test/java/ch/wisv/areafiftylan/integration/WebTournamentIntegrationTest.java]: Can't you simply fix this with a proper `equals()` method in `Tournament`?

File: src/test/java/ch/wisv/areafiftylan/integration/WebTournamentIntegrationTest.java [hunk 4]
```diff
@@ -212,4 +234,67 @@ public void testDeleteAllTournamentsAsCommitteeMember() {
...
+        assertEquals(0, tournamentRepository.findAll().size());
+
+        given()
+            .header(getXAuthTokenHeaderForUser(admin))
+            .body(updateTournament)
+            .contentType(ContentType.JSON)
+        .when()
+            .post(TOURNAMENT_ENDPOINT + updateTournament.getId())
+        .then()
+            .statusCode(HttpStatus.SC_INTERNAL_SERVER_ERROR)
```
Comment (skamoen) [inline src/test/java/ch/wisv/areafiftylan/integration/WebTournamentIntegrationTest.java]: This is not a server error, should be handled as a bad request or not found

Comment (Fastjur) [reply src/test/java/ch/wisv/areafiftylan/integration/WebTournamentIntegrationTest.java]: The `TournamentNotFound` exception extends  `AreaFiftyLANException` extends `RunTimeException`. Is there another way to do this?

File: src/test/java/ch/wisv/areafiftylan/integration/WebTournamentIntegrationTest.java (hunk 3 above)
Comment (Fastjur) [reply src/test/java/ch/wisv/areafiftylan/integration/WebTournamentIntegrationTest.java]: I could do that too, I was unsure if we want to write these ourselves. I'll do that :+1:

File: src/main/java/ch/wisv/areafiftylan/web/tournament/controller/TournamentController.java (hunk 2 above)
Comment (Fastjur) [reply src/main/java/ch/wisv/areafiftylan/web/tournament/controller/TournamentController.java]: Gonna be honest, I was pretty drunk when I wrote this :upside_down_face: 

I fixed it now ;)

File: src/main/java/ch/wisv/areafiftylan/web/tournament/model/Tournament.java [hunk 5]
```diff
@@ -3,6 +3,7 @@
 import ch.wisv.areafiftylan.web.sponsor.model.Sponsor;
 import lombok.Data;
 import lombok.NoArgsConstructor;
+import net.logstash.logback.encoder.org.apache.commons.lang.builder.EqualsBuilder;
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/web/tournament/model/Tournament.java]: What is this doing here?

File: src/main/java/ch/wisv/areafiftylan/web/tournament/model/Tournament.java [hunk 6]
```diff
@@ -32,4 +33,44 @@
...
+                if (!prizes.get(i).equals(that.prizes.get(i))) return false;
+            }
+        } else {
+            if (prizes == null ^ that.prizes == null) return false;
+        }
+        return sponsor != null ? sponsor.equals(that.sponsor) : that.sponsor == null;
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/web/tournament/model/Tournament.java]: Don't think you need this

File: src/main/java/ch/wisv/areafiftylan/web/tournament/controller/TournamentController.java [hunk 7]
```diff
@@ -42,6 +43,19 @@ public TournamentController(TournamentService tournamentService) {
...
+    @PutMapping("/{id}")
+    @PreAuthorize("hasRole('COMMITTEE')")
+    ResponseEntity<?> editTournament(@RequestBody Tournament tournament) {
+        final Tournament finalTournament = tournament;
+        tournamentService.getTournaments()
+                .stream()
+                .filter(dbTournament -> dbTournament.getId().equals(finalTournament.getId()))
+                .findFirst()
+                .orElseThrow(TournamentNotFoundException::new);
+        tournament = tournamentService.addTournament(tournament);
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/web/tournament/controller/TournamentController.java]: Please move all the logic to the tournamentService with a method `replaceTournament(id, tournament)` or something

**Comment by Fastjur** [comment]
@skamoen Codecov now fails on the equals method. Do we want to have 100% branch coverage on the equals method?

File: src/main/java/ch/wisv/areafiftylan/web/tournament/service/TournamentServiceImpl.java [hunk 8]
```diff
@@ -30,6 +31,17 @@ public Tournament addTournament(Tournament tournament) {
         return tournamentRepository.save(tournament);
     }
 
+    @Override
+    public Tournament replaceTournament(Tournament tournament) {
+        if (getTournaments()
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/web/tournament/service/TournamentServiceImpl.java]: So you're iterating over all tournaments, checking if none match the given ID? Why not just use [this](https://docs.spring.io/spring-data/jpa/docs/current/api/org/springframework/data/jpa/repository/JpaRepository.html#getOne-ID-)

**Pull request merged**